File: platform/polyglot/saucelabs_api/mjs/examples/basic-usage.mjs
/**
 * basic-usage.mjs — Sauce Labs API SDK Examples (Node.js)
 *
 * Self-contained demo script showcasing the core features of the SDK.
 * Each exampleX_name() function is independent and can be run in isolation.
 *
 * NOTE: This is a demonstration script. It sets fake credentials so that
 * client construction succeeds, and wraps all network-bound calls in
 * try/catch to illustrate the API surface without making real HTTP requests.
 *
 * Usage:
 *   node examples/basic-usage.mjs
 */

import {
  // Core
  SaucelabsClient,
  createSaucelabsClient,

  // Config
  resolveConfig,
  resolveCoreBaseUrl,
  resolveMobileBaseUrl,

  // Errors
  SaucelabsError,
  SaucelabsAuthError,
  SaucelabsNotFoundError,
  SaucelabsRateLimitError,
  SaucelabsValidationError,
  SaucelabsServerError,
  SaucelabsConfigError,
  createErrorFromResponse,

  // Rate limiting
  parseRetryAfter,
  buildRateLimitInfo,
  calculateBackoff,

  // Logger
  createLogger,

  // Types & constants
  DEFAULT_BASE_URL,
  DEFAULT_MOBILE_BASE_URL,
  DEFAULT_TIMEOUT,
  CORE_REGIONS,
  MOBILE_REGIONS,
  AUTOMATION_API_VALUES,

  // Domain modules
  JobsModule,
  PlatformModule,
  UsersModule,
  UploadModule,
} from '../src/index.mjs';

// Fake credentials used throughout examples so SaucelabsClient construction succeeds.
const DEMO_USERNAME = 'demo_user';
const DEMO_API_KEY = 'demo_access_key_00000000';


// =============================================================================
// Example 1: Client Initialization
// =============================================================================
/**
 * Demonstrates creating a SaucelabsClient with explicit credentials and
 * custom options. Prints the resolved configuration.
 */
function example1_clientInitialization() {
  console.log('\n--- Example 1: Client Initialization ---\n');

  const prevUser = process.env.SAUCE_USERNAME;
  const prevKey = process.env.SAUCE_ACCESS_KEY;
  process.env.SAUCE_USERNAME = DEMO_USERNAME;
  process.env.SAUCE_ACCESS_KEY = DEMO_API_KEY;
  process.env.LOG_LEVEL = 'silent';

  try {
    // Create a client with explicit options.
    const client = new SaucelabsClient({
      username: DEMO_USERNAME,
      apiKey: DEMO_API_KEY,
      region: 'us-west-1',
      timeout: 15000,
      rateLimitAutoWait: true,
    });

    console.log('Client created successfully.');
    console.log('  username:', client.username);
    console.log('  lastRateLimit:', client.lastRateLimit);

    client.close();
  } finally {
    if (prevUser === undefined) { delete process.env.SAUCE_USERNAME; } else { process.env.SAUCE_USERNAME = prevUser; }
    if (prevKey === undefined) { delete process.env.SAUCE_ACCESS_KEY; } else { process.env.SAUCE_ACCESS_KEY = prevKey; }
  }
}


// =============================================================================
// Example 2: Configuration Resolution
// =============================================================================
/**
 * Demonstrates resolveConfig(), resolveCoreBaseUrl(), and resolveMobileBaseUrl().
 * Shows how the SDK loads configuration from environment variables and explicit
 * overrides, and how different regions map to different API base URLs.
 */
function example2_configurationResolution() {
  console.log('\n--- Example 2: Configuration Resolution ---\n');

  const prevUser = process.env.SAUCE_USERNAME;
  const prevKey = process.env.SAUCE_ACCESS_KEY;
  process.env.SAUCE_USERNAME = DEMO_USERNAME;
  process.env.SAUCE_ACCESS_KEY = DEMO_API_KEY;
  process.env.LOG_LEVEL = 'silent';

  try {
    // 2a — Default configuration from environment.
    const defaultConfig = resolveConfig();
    console.log('Default config from env:');
    console.log('  username: ', defaultConfig.username);
    console.log('  baseUrl:  ', defaultConfig.baseUrl);
    console.log('  mobileUrl:', defaultConfig.mobileBaseUrl);
    console.log('  timeout:  ', defaultConfig.timeout, 'ms');
    console.log('  autoWait: ', defaultConfig.rateLimitAutoWait);
    console.log();

    // 2b — EU region override.
    const euConfig = resolveConfig({ region: 'eu-central-1' });
    console.log('EU region config:');
    console.log('  baseUrl:  ', euConfig.baseUrl);
    console.log('  mobileUrl:', euConfig.mobileBaseUrl);
    console.log();

    // 2c — Direct URL resolution.
    console.log('Core region URLs:');
    for (const [name, url] of Object.entries(CORE_REGIONS)) {
      console.log(`  ${name}: ${url}`);
    }
    console.log();

    console.log('Mobile region URLs:');
    for (const [name, url] of Object.entries(MOBILE_REGIONS)) {
      console.log(`  ${name}: ${url}`);
    }
    console.log();

    // 2d — Custom base URL override.
    const customUrl = resolveCoreBaseUrl('us-west-1', 'https://custom-proxy.example.com/');
    console.log('Custom base URL override:', customUrl);

    // 2e — SDK defaults.
    console.log('\nSDK defaults:');
    console.log('  DEFAULT_BASE_URL:       ', DEFAULT_BASE_URL);
    console.log('  DEFAULT_MOBILE_BASE_URL:', DEFAULT_MOBILE_BASE_URL);
    console.log('  DEFAULT_TIMEOUT:        ', DEFAULT_TIMEOUT, 'ms');
  } finally {
    if (prevUser === undefined) { delete process.env.SAUCE_USERNAME; } else { process.env.SAUCE_USERNAME = prevUser; }
    if (prevKey === undefined) { delete process.env.SAUCE_ACCESS_KEY; } else { process.env.SAUCE_ACCESS_KEY = prevKey; }
  }
}


// =============================================================================
// Example 3: Error Handling
// =============================================================================
/**
 * Demonstrates the typed error hierarchy. Each error class is instantiated
 * directly to show its properties and the toJSON() serialization format.
 * In production these are thrown by the SDK when the API returns errors.
 */
async function example3_errorHandling() {
  console.log('\n--- Example 3: Error Handling ---\n');

  // 3a — Demonstrate each error type.
  const errors = [
    new SaucelabsAuthError('Invalid credentials', { statusCode: 401 }),
    new SaucelabsNotFoundError('Job abc123 not found', { statusCode: 404 }),
    new SaucelabsRateLimitError('Rate limit exceeded', { retryAfter: 30 }),
    new SaucelabsValidationError('Invalid parameter: limit must be > 0', { statusCode: 400 }),
    new SaucelabsServerError('Internal server error', { statusCode: 500 }),
    new SaucelabsConfigError('Missing SAUCE_USERNAME environment variable'),
  ];

  for (const err of errors) {
    console.log(`${err.name}:`);
    console.log('  message:   ', err.message);
    console.log('  statusCode:', err.statusCode);
    console.log('  toJSON():  ', JSON.stringify(err.toJSON(), null, 4));
    console.log('  instanceof SaucelabsError:', err instanceof SaucelabsError);
    console.log();
  }

  // 3b — Demonstrate createErrorFromResponse factory.
  console.log('-- createErrorFromResponse factory --\n');

  const err401 = createErrorFromResponse(401, { message: 'Unauthorized' }, {});
  console.log('401 ->', err401.name, ':', err401.message);

  const err429 = createErrorFromResponse(429, { message: 'Too many requests' }, { 'retry-after': '60' });
  console.log('429 ->', err429.name, ': retryAfter =', err429.retryAfter, 's');

  const err500 = createErrorFromResponse(500, { message: 'Server error' }, {});
  console.log('500 ->', err500.name, ':', err500.message);
  console.log();

  // 3c — Try/catch pattern with a real client.
  process.env.SAUCE_USERNAME = DEMO_USERNAME;
  process.env.SAUCE_ACCESS_KEY = DEMO_API_KEY;
  process.env.LOG_LEVEL = 'silent';

  const client = new SaucelabsClient({
    username: DEMO_USERNAME,
    apiKey: DEMO_API_KEY,
    timeout: 3000,
  });

  try {
    console.log('-- Simulated client.get() error catch pattern --\n');
    await client.get('/rest/v1/demo_user/jobs');
  } catch (err) {
    if (err instanceof SaucelabsAuthError) {
      console.log('  Caught SaucelabsAuthError:', err.message);
    } else if (err instanceof SaucelabsNotFoundError) {
      console.log('  Caught SaucelabsNotFoundError:', err.message);
    } else if (err instanceof SaucelabsRateLimitError) {
      console.log('  Caught SaucelabsRateLimitError: retry after', err.retryAfter, 's');
    } else if (err instanceof SaucelabsError) {
      console.log('  Caught SaucelabsError:', err.name, '-', err.message);
    } else {
      console.log('  Caught unexpected error (network expected in demo):', err.message);
    }
  } finally {
    client.close();
  }
}


// =============================================================================
// Example 4: Convenience Factory
// =============================================================================
/**
 * Demonstrates createSaucelabsClient() which creates a SaucelabsClient with
 * all domain modules (jobs, platform, users, upload) pre-attached.
 */
function example4_convenienceFactory() {
  console.log('\n--- Example 4: Convenience Factory ---\n');

  process.env.SAUCE_USERNAME = DEMO_USERNAME;
  process.env.SAUCE_ACCESS_KEY = DEMO_API_KEY;
  process.env.LOG_LEVEL = 'silent';

  const client = createSaucelabsClient({
    username: DEMO_USERNAME,
    apiKey: DEMO_API_KEY,
    region: 'us-west-1',
  });

  console.log('Client created via createSaucelabsClient():');
  console.log('  username:    ', client.username);
  console.log('  client.jobs: ', client.jobs instanceof JobsModule);
  console.log('  client.platform:', client.platform instanceof PlatformModule);
  console.log('  client.users:', client.users instanceof UsersModule);
  console.log('  client.upload:', client.upload instanceof UploadModule);

  client.close();
}


// =============================================================================
// Example 5: Domain Modules
// =============================================================================
/**
 * Demonstrates creating domain modules (JobsModule, PlatformModule,
 * UsersModule) and invoking their methods. Since there is no real API,
 * each call is wrapped in try/catch.
 */
async function example5_domainModules() {
  console.log('\n--- Example 5: Domain Modules ---\n');

  process.env.SAUCE_USERNAME = DEMO_USERNAME;
  process.env.SAUCE_ACCESS_KEY = DEMO_API_KEY;
  process.env.LOG_LEVEL = 'silent';

  const client = createSaucelabsClient({
    username: DEMO_USERNAME,
    apiKey: DEMO_API_KEY,
    timeout: 3000,
  });

  // 5a — JobsModule: list jobs
  try {
    console.log('Calling client.jobs.list({ limit: 5 }) ...');
    await client.jobs.list({ limit: 5 });
  } catch (err) {
    console.log('  Expected error (no real API):', err.constructor.name, '-', err.message, '\n');
  }

  // 5b — JobsModule: get a specific job
  try {
    console.log('Calling client.jobs.get("abc123def456") ...');
    await client.jobs.get('abc123def456');
  } catch (err) {
    console.log('  Expected error:', err.constructor.name, '-', err.message, '\n');
  }

  // 5c — PlatformModule: get service status (public, no auth needed)
  try {
    console.log('Calling client.platform.getStatus() ...');
    await client.platform.getStatus();
  } catch (err) {
    console.log('  Expected error:', err.constructor.name, '-', err.message, '\n');
  }

  // 5d — PlatformModule: get supported platforms
  try {
    console.log('Calling client.platform.getPlatforms("appium") ...');
    await client.platform.getPlatforms('appium');
  } catch (err) {
    console.log('  Expected error:', err.constructor.name, '-', err.message, '\n');
  }

  // 5e — UsersModule: get user info
  try {
    console.log('Calling client.users.getUser() ...');
    await client.users.getUser();
  } catch (err) {
    console.log('  Expected error:', err.constructor.name, '-', err.message, '\n');
  }

  // 5f — UsersModule: get concurrency
  try {
    console.log('Calling client.users.getConcurrency() ...');
    await client.users.getConcurrency();
  } catch (err) {
    console.log('  Expected error:', err.constructor.name, '-', err.message, '\n');
  }

  // 5g — Validation: invalid automation API
  try {
    console.log('Calling client.platform.getPlatforms("invalid") ...');
    await client.platform.getPlatforms('invalid');
  } catch (err) {
    console.log('  Validation error (expected):', err.constructor.name, '-', err.message, '\n');
  }

  console.log('Valid automation API values:', AUTOMATION_API_VALUES.join(', '));

  client.close();
}


// =============================================================================
// Example 6: Rate Limit Handling
// =============================================================================
/**
 * Demonstrates rate limit utilities: parseRetryAfter for extracting wait
 * times from headers, buildRateLimitInfo for constructing rate limit state,
 * and calculateBackoff for exponential backoff with jitter.
 */
function example6_rateLimitHandling() {
  console.log('\n--- Example 6: Rate Limit Handling ---\n');

  // 6a — parseRetryAfter with various inputs.
  console.log('-- parseRetryAfter --\n');
  console.log('  parseRetryAfter("30"):    ', parseRetryAfter('30'));
  console.log('  parseRetryAfter("1.5"):   ', parseRetryAfter('1.5'));
  console.log('  parseRetryAfter(null):    ', parseRetryAfter(null));
  console.log('  parseRetryAfter(""):      ', parseRetryAfter(''));
  console.log();

  // 6b — buildRateLimitInfo with mock 429 response headers.
  console.log('-- buildRateLimitInfo --\n');
  const mockHeaders = {
    'retry-after': '30',
    'x-ratelimit-remaining': '0',
    'x-ratelimit-limit': '100',
    'x-ratelimit-reset': String(Math.floor(Date.now() / 1000) + 60),
  };
  const info = buildRateLimitInfo(mockHeaders);
  console.log('  Mock headers:', JSON.stringify(mockHeaders, null, 2));
  console.log('  Rate limit info:');
  console.log('    retryAfter:', info.retryAfter);
  console.log('    remaining: ', info.remaining);
  console.log('    limit:     ', info.limit);
  console.log('    resetAt:   ', info.resetAt);
  console.log('    timestamp: ', info.timestamp);
  console.log();

  // 6c — calculateBackoff for exponential backoff.
  console.log('-- calculateBackoff (base=1s, max=60s) --\n');
  for (let attempt = 0; attempt < 6; attempt++) {
    const delay = calculateBackoff(attempt, 1, 60);
    console.log(`  Attempt ${attempt}: ~${delay.toFixed(1)}s`);
  }
  console.log();

  // 6d — Client with onRateLimit callback.
  process.env.SAUCE_USERNAME = DEMO_USERNAME;
  process.env.SAUCE_ACCESS_KEY = DEMO_API_KEY;
  process.env.LOG_LEVEL = 'silent';

  const client = new SaucelabsClient({
    username: DEMO_USERNAME,
    apiKey: DEMO_API_KEY,
    rateLimitAutoWait: true,
    onRateLimit: (info) => {
      console.log('  [onRateLimit callback] retryAfter:', info.retryAfter);
      return true; // allow auto-wait
    },
  });
  console.log('Client created with onRateLimit callback and rateLimitAutoWait=true.');
  client.close();
}


// =============================================================================
// Example 7: Logging
// =============================================================================
/**
 * Demonstrates the SDK logger with different levels and custom loggers.
 * Shows how sensitive data is automatically redacted.
 */
function example7_logging() {
  console.log('\n--- Example 7: Logging ---\n');

  // 7a — Create a logger with debug level.
  process.env.LOG_LEVEL = 'debug';
  const logger = createLogger('saucelabs-example', 'basic-usage');
  logger.info('this is an info message');
  logger.debug('this is a debug message');
  logger.warn('this is a warning');
  logger.error('this is an error message');
  console.log();

  // 7b — Custom logger.
  const logs = [];
  const customLogger = {
    debug: (msg) => logs.push(`[DEBUG] ${msg}`),
    info: (msg) => logs.push(`[INFO] ${msg}`),
    warn: (msg) => logs.push(`[WARN] ${msg}`),
    error: (msg) => logs.push(`[ERROR] ${msg}`),
  };

  process.env.LOG_LEVEL = 'silent';
  const client = new SaucelabsClient({
    username: DEMO_USERNAME,
    apiKey: DEMO_API_KEY,
    logger: customLogger,
  });

  console.log('Custom logger captured', logs.length, 'messages:');
  for (const line of logs) {
    console.log(' ', line);
  }

  client.close();
}


// =============================================================================
// Main — Run all examples sequentially
// =============================================================================
async function main() {
  console.log('='.repeat(72));
  console.log(' Sauce Labs API SDK — Basic Usage Examples (Node.js / MJS)');
  console.log('='.repeat(72));

  try { example1_clientInitialization(); } catch (err) { console.error('Example 1 failed:', err.message); }
  try { example2_configurationResolution(); } catch (err) { console.error('Example 2 failed:', err.message); }
  try { await example3_errorHandling(); } catch (err) { console.error('Example 3 failed:', err.message); }
  try { example4_convenienceFactory(); } catch (err) { console.error('Example 4 failed:', err.message); }
  try { await example5_domainModules(); } catch (err) { console.error('Example 5 failed:', err.message); }
  try { example6_rateLimitHandling(); } catch (err) { console.error('Example 6 failed:', err.message); }
  try { example7_logging(); } catch (err) { console.error('Example 7 failed:', err.message); }

  console.log('\n' + '='.repeat(72));
  console.log(' All examples complete.');
  console.log('='.repeat(72));
}

main().catch((err) => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: platform/polyglot/saucelabs_api/py/examples/basic_usage.py
"""
Basic Usage Examples — Sauce Labs API SDK (Python)

Self-contained demonstration of all core SDK features.
Each exampleX_name() function is independent and shows one concept.

Usage:
    python basic_usage.py

Prerequisites:
    pip install -e ".[dev]"    (from the py/ directory)
"""

import asyncio
import json
import os

from saucelabs_api import (
    # Core
    SaucelabsClient,
    create_saucelabs_client,
    # Config
    resolve_config,
    resolve_core_base_url,
    resolve_mobile_base_url,
    # Errors
    SaucelabsError,
    SaucelabsAuthError,
    SaucelabsNotFoundError,
    SaucelabsRateLimitError,
    SaucelabsValidationError,
    SaucelabsServerError,
    SaucelabsConfigError,
    create_error_from_response,
    # Rate limiter
    RateLimiter,
    # Logger
    create_logger,
    # Types & constants
    DEFAULT_BASE_URL,
    DEFAULT_MOBILE_BASE_URL,
    DEFAULT_TIMEOUT,
    CORE_REGIONS,
    MOBILE_REGIONS,
    AUTOMATION_API_VALUES,
    # Domain modules
    JobsModule,
    PlatformModule,
    UsersModule,
    UploadModule,
)

# Fake credentials used throughout examples so client construction succeeds.
# This is NOT a real account — HTTP requests will fail, which we catch gracefully.
DEMO_USERNAME = "demo_user"
DEMO_API_KEY = "demo_access_key_00000000"


def _separator(title: str) -> None:
    """Print a visual separator between examples."""
    print(f"\n{'=' * 72}")
    print(f"  {title}")
    print(f"{'=' * 72}\n")


# =============================================================================
# Example 1: Client Initialization
# =============================================================================
async def example1_client_initialization() -> None:
    """
    Create a SaucelabsClient with explicit credentials and custom options.
    Uses the async context manager pattern for automatic cleanup.
    """
    _separator("Example 1: Client Initialization")

    os.environ["SAUCE_USERNAME"] = DEMO_USERNAME
    os.environ["SAUCE_ACCESS_KEY"] = DEMO_API_KEY
    os.environ["LOG_LEVEL"] = "SILENT"

    async with SaucelabsClient(
        username=DEMO_USERNAME,
        api_key=DEMO_API_KEY,
        region="us-west-1",
        timeout=15.0,
        rate_limit_auto_wait=True,
    ) as client:
        print("Client created successfully with custom options:")
        print(f"  username:       {client.username}")
        print(f"  lastRateLimit:  {client.last_rate_limit}")

    del os.environ["SAUCE_USERNAME"]
    del os.environ["SAUCE_ACCESS_KEY"]
    print("\n[OK] Client initialization complete.")


# =============================================================================
# Example 2: Configuration Resolution
# =============================================================================
def example2_configuration_resolution() -> None:
    """
    Demonstrate resolve_config(), resolve_core_base_url(), and
    resolve_mobile_base_url(). Shows how different regions map to
    different API base URLs and how environment variables are resolved.
    """
    _separator("Example 2: Configuration Resolution")

    os.environ["SAUCE_USERNAME"] = DEMO_USERNAME
    os.environ["SAUCE_ACCESS_KEY"] = DEMO_API_KEY
    os.environ["LOG_LEVEL"] = "SILENT"

    # 2a — Default configuration from environment.
    config = resolve_config()
    print("Default config from env:")
    print(f"  username:       {config['username']}")
    print(f"  base_url:       {config['base_url']}")
    print(f"  mobile_base_url:{config['mobile_base_url']}")
    print(f"  timeout:        {config['timeout']}s")
    print(f"  auto_wait:      {config['rate_limit_auto_wait']}")
    print()

    # 2b — EU region override.
    eu_config = resolve_config(region="eu-central-1")
    print("EU region config:")
    print(f"  base_url:       {eu_config['base_url']}")
    print(f"  mobile_base_url:{eu_config['mobile_base_url']}")
    print()

    # 2c — Direct URL resolution.
    print("Core region URLs:")
    for name, url in CORE_REGIONS.items():
        print(f"  {name}: {url}")
    print()

    print("Mobile region URLs:")
    for name, url in MOBILE_REGIONS.items():
        print(f"  {name}: {url}")
    print()

    # 2d — Custom base URL override.
    custom_url = resolve_core_base_url("us-west-1", "https://custom-proxy.example.com/")
    print(f"Custom base URL override: {custom_url}")

    # 2e — SDK defaults.
    print("\nSDK defaults:")
    print(f"  DEFAULT_BASE_URL:        {DEFAULT_BASE_URL}")
    print(f"  DEFAULT_MOBILE_BASE_URL: {DEFAULT_MOBILE_BASE_URL}")
    print(f"  DEFAULT_TIMEOUT:         {DEFAULT_TIMEOUT}s")

    del os.environ["SAUCE_USERNAME"]
    del os.environ["SAUCE_ACCESS_KEY"]


# =============================================================================
# Example 3: Error Handling
# =============================================================================
async def example3_error_handling() -> None:
    """
    Demonstrate the typed error hierarchy and create_error_from_response
    factory. Shows the try/except pattern for client operations.
    """
    _separator("Example 3: Error Handling")

    # 3a — Demonstrate each error type.
    errors = [
        SaucelabsAuthError("Invalid credentials", status_code=401),
        SaucelabsNotFoundError("Job abc123 not found", status_code=404),
        SaucelabsRateLimitError("Rate limit exceeded", retry_after=30.0),
        SaucelabsValidationError("Invalid parameter", status_code=400),
        SaucelabsServerError("Internal server error", status_code=500),
        SaucelabsConfigError("Missing SAUCE_USERNAME"),
    ]

    for err in errors:
        print(f"{type(err).__name__}:")
        print(f"  message:     {err}")
        print(f"  status_code: {err.status_code}")
        print(f"  isinstance(SaucelabsError): {isinstance(err, SaucelabsError)}")
        print()

    # 3b — create_error_from_response factory.
    print("-- create_error_from_response factory --\n")

    err_401 = create_error_from_response(401, {"message": "Unauthorized"}, {})
    print(f"401 -> {type(err_401).__name__}: {err_401}")

    err_429 = create_error_from_response(429, {"message": "Too many requests"}, {"retry-after": "60"})
    print(f"429 -> {type(err_429).__name__}: retry_after = {err_429.retry_after}s")

    err_500 = create_error_from_response(500, {"message": "Server error"}, {})
    print(f"500 -> {type(err_500).__name__}: {err_500}")
    print()

    # 3c — Try/except pattern with a real client.
    os.environ["SAUCE_USERNAME"] = DEMO_USERNAME
    os.environ["SAUCE_ACCESS_KEY"] = DEMO_API_KEY
    os.environ["LOG_LEVEL"] = "SILENT"

    print("-- Simulated client.get() error catch pattern --\n")
    async with SaucelabsClient(
        username=DEMO_USERNAME,
        api_key=DEMO_API_KEY,
        timeout=3.0,
    ) as client:
        try:
            await client.get("/rest/v1/demo_user/jobs")
        except SaucelabsAuthError as err:
            print(f"  Caught SaucelabsAuthError: {err}")
        except SaucelabsNotFoundError as err:
            print(f"  Caught SaucelabsNotFoundError: {err}")
        except SaucelabsRateLimitError as err:
            print(f"  Caught SaucelabsRateLimitError: retry_after={err.retry_after}s")
        except SaucelabsError as err:
            print(f"  Caught SaucelabsError ({type(err).__name__}): {err}")
        except Exception as err:
            print(f"  Caught unexpected error (network expected in demo): {type(err).__name__}: {err}")

    del os.environ["SAUCE_USERNAME"]
    del os.environ["SAUCE_ACCESS_KEY"]
    print("\n[OK] Error handling examples complete.")


# =============================================================================
# Example 4: Convenience Factory
# =============================================================================
def example4_convenience_factory() -> None:
    """
    Demonstrate create_saucelabs_client() which creates a SaucelabsClient
    with all domain modules (jobs, platform, users, upload) pre-attached.
    """
    _separator("Example 4: Convenience Factory")

    os.environ["SAUCE_USERNAME"] = DEMO_USERNAME
    os.environ["SAUCE_ACCESS_KEY"] = DEMO_API_KEY
    os.environ["LOG_LEVEL"] = "SILENT"

    client = create_saucelabs_client(
        username=DEMO_USERNAME,
        api_key=DEMO_API_KEY,
        region="us-west-1",
    )

    print("Client created via create_saucelabs_client():")
    print(f"  username:         {client.username}")
    print(f"  client.jobs:      {isinstance(client.jobs, JobsModule)}")
    print(f"  client.platform:  {isinstance(client.platform, PlatformModule)}")
    print(f"  client.users:     {isinstance(client.users, UsersModule)}")
    print(f"  client.upload:    {isinstance(client.upload, UploadModule)}")

    del os.environ["SAUCE_USERNAME"]
    del os.environ["SAUCE_ACCESS_KEY"]


# =============================================================================
# Example 5: Domain Modules
# =============================================================================
async def example5_domain_modules() -> None:
    """
    Demonstrate domain modules (JobsModule, PlatformModule, UsersModule)
    and their method signatures. Since there is no real API, each call is
    wrapped in try/except.
    """
    _separator("Example 5: Domain Modules")

    os.environ["SAUCE_USERNAME"] = DEMO_USERNAME
    os.environ["SAUCE_ACCESS_KEY"] = DEMO_API_KEY
    os.environ["LOG_LEVEL"] = "SILENT"

    client = create_saucelabs_client(
        username=DEMO_USERNAME,
        api_key=DEMO_API_KEY,
        timeout=3.0,
    )

    async with client:
        # 5a — JobsModule: list jobs
        try:
            print("Calling client.jobs.list({'limit': 5}) ...")
            await client.jobs.list(params={"limit": 5})
        except Exception as err:
            print(f"  Expected error (no real API): {type(err).__name__}: {err}\n")

        # 5b — JobsModule: get a specific job
        try:
            print('Calling client.jobs.get("abc123def456") ...')
            await client.jobs.get("abc123def456")
        except Exception as err:
            print(f"  Expected error: {type(err).__name__}: {err}\n")

        # 5c — PlatformModule: get service status (public, no auth needed)
        try:
            print("Calling client.platform.get_status() ...")
            await client.platform.get_status()
        except Exception as err:
            print(f"  Expected error: {type(err).__name__}: {err}\n")

        # 5d — PlatformModule: get supported platforms
        try:
            print('Calling client.platform.get_platforms("appium") ...')
            await client.platform.get_platforms("appium")
        except Exception as err:
            print(f"  Expected error: {type(err).__name__}: {err}\n")

        # 5e — UsersModule: get user info
        try:
            print("Calling client.users.get_user() ...")
            await client.users.get_user()
        except Exception as err:
            print(f"  Expected error: {type(err).__name__}: {err}\n")

        # 5f — UsersModule: get concurrency
        try:
            print("Calling client.users.get_concurrency() ...")
            await client.users.get_concurrency()
        except Exception as err:
            print(f"  Expected error: {type(err).__name__}: {err}\n")

        # 5g — Validation: invalid automation API
        try:
            print('Calling client.platform.get_platforms("invalid") ...')
            await client.platform.get_platforms("invalid")
        except SaucelabsValidationError as err:
            print(f"  Validation error (expected): {err}\n")
        except Exception as err:
            print(f"  Expected error: {type(err).__name__}: {err}\n")

    print(f"Valid automation API values: {', '.join(AUTOMATION_API_VALUES)}")

    del os.environ["SAUCE_USERNAME"]
    del os.environ["SAUCE_ACCESS_KEY"]


# =============================================================================
# Example 6: Rate Limit Handling
# =============================================================================
def example6_rate_limit_handling() -> None:
    """
    Demonstrate rate limit utilities and callback configuration.
    Shows how exponential backoff works and how to handle 429 responses.
    """
    _separator("Example 6: Rate Limit Handling")

    from saucelabs_api.errors import _parse_retry_after

    # 6a — Parse retry-after headers.
    print("-- _parse_retry_after --\n")
    print(f"  _parse_retry_after({{'retry-after': '30'}}):  {_parse_retry_after({'retry-after': '30'})}")
    print(f"  _parse_retry_after({{'retry-after': '1.5'}}): {_parse_retry_after({'retry-after': '1.5'})}")
    print(f"  _parse_retry_after({{}}):                     {_parse_retry_after({})}")
    print()

    # 6b — Exponential backoff.
    print("-- RateLimiter._calculate_backoff (base=1.0, max=60.0) --\n")
    for attempt in range(6):
        delay = RateLimiter._calculate_backoff(attempt, base_delay=1.0, max_delay=60.0)
        print(f"  Attempt {attempt}: ~{delay:.1f}s")
    print()

    # 6c — RateLimiter with callback.
    os.environ["LOG_LEVEL"] = "SILENT"
    logger = create_logger("example", "rate-limit")

    rl = RateLimiter(
        auto_wait=True,
        max_retries=5,
        on_rate_limit=lambda info: print(f"  [callback] Rate limited! retryAfter={info.retry_after}"),
        logger=logger,
    )
    print("RateLimiter created:")
    print(f"  last_rate_limit: {rl.last_rate_limit}")
    print("  auto_wait: True, max_retries: 5")


# =============================================================================
# Example 7: Logging
# =============================================================================
def example7_logging() -> None:
    """
    Demonstrate the SDK logger with different levels and sensitive data
    redaction.
    """
    _separator("Example 7: Logging")

    # 7a — Create a logger with debug level.
    os.environ["LOG_LEVEL"] = "DEBUG"
    logger = create_logger("saucelabs-example", "basic-usage")
    logger.info("this is an info message")
    logger.debug("this is a debug message")
    logger.warning("this is a warning")
    logger.error("this is an error message")
    print()

    # 7b — Redaction of sensitive data.
    from saucelabs_api.logger import _redact_context

    context = {
        "username": "demo_user",
        "access_key": "super_secret_key_12345",
        "api_key": "another_secret",
        "region": "us-west-1",
        "token": "tok_abc123",
    }
    redacted = _redact_context(context)
    print("Sensitive data redaction:")
    print(f"  Original: {json.dumps(context, indent=4)}")
    print(f"  Redacted: {json.dumps(redacted, indent=4)}")

    os.environ["LOG_LEVEL"] = "SILENT"


# =============================================================================
# Main — Run all examples sequentially
# =============================================================================
async def main() -> None:
    """Run all example functions in order."""
    print("=" * 72)
    print("  Sauce Labs API SDK — Python Examples")
    print("=" * 72)
    print()
    print("This script demonstrates all core features of the saucelabs_api SDK.")
    print("Fake demo credentials are used — HTTP requests will fail gracefully.")
    print()

    await example1_client_initialization()
    example2_configuration_resolution()
    await example3_error_handling()
    example4_convenience_factory()
    await example5_domain_modules()
    example6_rate_limit_handling()
    example7_logging()

    print()
    print("=" * 72)
    print("  All examples completed successfully.")
    print("=" * 72)


if __name__ == "__main__":
    asyncio.run(main())
